icu4c/source/i18n/usearch.cpp

```diff
@@ -187,21 +188,24 @@ inline void * allocateMemory(uint32_t size, UErrorCode *status)
 */
 static
 inline int32_t * addTouint32_tArray(int32_t    *destination,
-                                    uint32_t    offset,
                                     uint32_t   *destinationlength,
+                                    bool        destOnHeap,
+                                    uint32_t    offset,
                                     uint32_t    value,
                                     uint32_t    increments,
                                     UErrorCode *status)
 {
-    uint32_t newlength = *destinationlength;
-    if (offset + 1 == newlength) {
-        newlength += increments;
+    if (offset >= *destinationlength) {
+        uint32_t newlength = offset + increments;
```

Comment:
The *status could be either success or failure when calling into addTouint32_tArray, the code does not check the status and call allocateMemory. And allocateMemory call uprv_malloc unconditionally. So this could cause a possible memory leak here in the if (U_FAILURE(*status)) { return }  if addTouint32_tArray is called with a failed status code. This is not caused by the change here but a preexisting issue. 
I think we should add 
if (U_FAILURE(*status)) { return destination; } 
in the beginning of this function.
or, if not desirable, add
U_ASSSERT(U_SUCCESS(*status));

Reply:
please look at the implementation of allocateMemory
it call uprv_malloc() without checking status. I think that is the major problem.

Reply:
Good catch. You are right, I didn't guard my new calls with status checks.
I changed these functions, inlining the allocation and always checking first for a failure code, as usual, so that no one gets confused.